aten/src/ATen/native/Linear.cpp

```diff
@@ -641,9 +642,26 @@ Tensor tensordot(const Tensor& input1, const Tensor& input2, IntArrayRef dims1,
 }
 
 Tensor &tensordot_out(const Tensor& input1, const Tensor& input2, IntArrayRef dims1, IntArrayRef dims2, Tensor& result) {
-  result.copy_(at::native::tensordot(input1, input2, dims1, dims2));
+  Tensor result_tmp = at::native::tensordot(input1, input2, dims1, dims2);
+  auto result_dtype = result_tmp.scalar_type();
+  auto output_tensor_dtype = result.scalar_type();
+  auto output_device = result.device();
+  auto input_device = input1.device();
+  // check if the input & output tensors are on the same device.
+  TORCH_CHECK(
+    output_device == input_device,
+    "tensordot_out: Expected the output and input tensors to be on the "
+    "same device, but got output on ", output_device, " and inputs on ",
+    input_device);
+  // check if the computed result can be safely copied to the output tensor.
+  bool can_cast = c10::canCast(result_dtype, output_tensor_dtype);
+  TORCH_CHECK(
+    can_cast, "tensordot_out",
+    ": Expected the output tensor to be safely castable from ", result_dtype,
+    " dtype, but got output tensor with dtype ", output_tensor_dtype);
+  at::native::resize_output(result, result_tmp.sizes());
+  result.copy_(result_tmp);
```

Comment:
Cool; note that here the use of out= provides no memory benefit, but this is not a regression.

Reply:
Thanks! I hadn't given much thought to this nuance. For my own edification, I'll see if some other ops handle things differently.

Reply:
Ideally the tensor provided to out=, if it's the correct size, can be reused during the computation to prevent other memory allocations. This is not always possible however, and it's probably not worth pursuing as there are more critical bugs and optimizations to address.

Reply:
Thanks for the clarification! I did wonder why there wasn't any conditional statement to match the size of the tensor provided to out= with that of the result, even for some other ops in `LinearAlgebra.cpp`. I'll take a look again & if some other operations are using such a conditional, I can reuse it wherever it can be used, or write a function to do so, probably in another PR..